File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        //Dynamic programming solution inspired by House Robber.
        //Accumulate the max at our current point, and then see if anything can beat it.
        //Assume that we hold onto our lowest for possible selling at any time.
        //O(n) runtime due to only 1 iteration through the list, O(n) space due to holding all of the max.
        //We can further reduce space by only holding onto the most recent (and last) profits.
        // int[] max = new int[prices.length]; //Further optimized.
        int previousMax = 0;
        int lowest = prices[0];
        for(int i = 1; i < prices.length; i++) {
            previousMax = Math.max(prices[i] - lowest, previousMax);
            if(prices[i] < lowest) lowest = prices[i];
        }
        return previousMax;
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.kt
class Solution {
    fun maxProfit(prices: IntArray): Int {
        //This is a dynamic programming kind of problem, as evidenced by the fact that we have overlapping subproblems (profit depends on when you bought in) and the optimal substructure (our best profits will exist across the entire problem.)
        //Start by declaring a current max array, set to 0, equal in size to our input.
        var max = 0 //Minor space + runtime optimization - we don't need the information from all days, only the most recent one.
        var localMin = prices[0] //Our lowest value so far, which we will use to check the profit.
        prices.forEach { value ->
            if(max < value - localMin) {
                max = value - localMin
            }
            if(value < localMin) localMin = value //Set the new low for future items
                //Optimal substructure here: If a new min exists, then any value after would achieve a higher profit by buying now and selling later vs. buying before and selling then.
            }
        return max
    }
}